src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/MessagePackHubProtocol.java

```diff
@@ -0,0 +1,572 @@
+        boolean allowReconnect = false;
+        
+        if (itemCount > 2) {
+            allowReconnect = unpacker.unpackBoolean();
+        }
+        
+        return new CloseMessage(error, allowReconnect);
+    }
+    
+    private String writeInvocationMessage(InvocationMessage message) throws IOException {
+        MessageBufferPacker packer = MessagePack.newDefaultBufferPacker();
+        
+        packer.packArrayHeader(6);
+        packer.packInt(message.getMessageType().value);
+        
+        writeHeaders(message.getHeaders(), packer);
+        
+        String invocationId = message.getInvocationId();
+        if (invocationId != null && !invocationId.isEmpty()) {
+            packer.packString(invocationId);
+        } else {
+            packer.packNil();
+        }
+        
+        packer.packString(message.getTarget());
+        
+        Object[] arguments = message.getArguments();
+        packer.packArrayHeader(arguments.length);
+        
+        for (Object o: arguments) {
+            writeValue(o, packer);
+        }
+        
+        writeStreamIds(message.getStreamIds(), packer);
+        
+        packer.flush();
+        String content = new String(packer.toByteArray(), StandardCharsets.ISO_8859_1);
+        packer.close();
+        return content;
+    }
+    
+    private String writeStreamItemMessage(StreamItem message) throws IOException {
+        MessageBufferPacker packer = MessagePack.newDefaultBufferPacker();
+        
+        packer.packArrayHeader(4);
+        packer.packInt(message.getMessageType().value);
+        
+        writeHeaders(message.getHeaders(), packer);
+        
+        packer.packString(message.getInvocationId());
+        
+        writeValue(message.getItem(), packer);
+        
+        packer.flush();
+        String content = new String(packer.toByteArray(), StandardCharsets.ISO_8859_1);
+        packer.close();
+        return content;
+    }
+    
+    private String writeCompletionMessage(CompletionMessage message) throws IOException {
+        MessageBufferPacker packer = MessagePack.newDefaultBufferPacker();
+        int resultKind =
+            message.getError() != null ? ERROR_RESULT :
+            message.getResult() != null ? NON_VOID_RESULT :
+            VOID_RESULT;
+        
+        packer.packArrayHeader(4 + (resultKind != VOID_RESULT ? 1: 0));
+        packer.packInt(message.getMessageType().value);
+        
+        writeHeaders(message.getHeaders(), packer);
+        
+        packer.packString(message.getInvocationId());
+        packer.packInt(resultKind);
+        
+        switch (resultKind) {
+        case ERROR_RESULT:
+            packer.packString(message.getError());
+            break;
+        case NON_VOID_RESULT:
+            writeValue(message.getResult(), packer);
+            break;
+        }
+        
+        packer.flush();
+        String content = new String(packer.toByteArray(), StandardCharsets.ISO_8859_1);
```

Comment:
Trust me, this is totally latin1 encoded... 😆

I think now is the time to change HubProtocol.writeMessage to return a ByteBuffer. AFAICT, HubProtocol is also package-private, so this should be NBD.

It also looks like MessageBufferPacker has a easy and efficient way to get a ByteBuffer via `.toMessageBuffer().sliceAsByteBuffer()`.

Reply:
Question about this one - changing `writeMessage` to return a `ByteBuffer`, and `parseMessage` to take the same, certainly makes things easier for messagePack, but doesn't it present the inverse of the same problem for JSON? Instead of MessagePack having to convert to and from Strings/byte arrays, now Json will have to do the same. @BrennanConroy @halter73

Reply:
Yes, but the JSON payloads start out as bytes that are implicitly interpreted as a UTF-8 string by the HttpClient. We should just do this explicitly in JSON protocol. This doesn't create any more work for the JSON protocol overall, it just moves where the work is done.

Pretending that the MessagePack buffers are really latin1-encoded string does work, but it's a hack that leads to unnecessary string allocations like the one this thread is about.

I think one of the problems here is that our HttpResponse.getContent() method returns a String instead of a ByteArray, but we can change that. Our HttpClient abstraction has never been public, so it's not breaking to change it. And OkHttp definitely supports getting a ByteArray instead of String representation of the body.

For example we could change `.string()` to `.bytes()` in the following line to get a ByteArray instead of a String (see https://square.github.io/okhttp/4.x/okhttp/okhttp3/-response-body/bytes/): 

https://github.com/dotnet/aspnetcore/blob/3192da444333cfbc30e4aff33a748dea76363756/src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/DefaultHttpClient.java#L152

And the OkHttp RequestBody type must support bytes because we use an empty byte array when sending an empty request body here:

https://github.com/dotnet/aspnetcore/blob/3192da444333cfbc30e4aff33a748dea76363756/src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/DefaultHttpClient.java#L115-L119

I can see changing the HttpClient abstraction as being kinda big, so I'm fine with doing this in a follow-up PR